R/cfr_time_varying.R

```diff
@@ -0,0 +1,229 @@
+  # check that input `<data.frame>` has columns date, cases, and deaths
+  checkmate::assert_names(
+    colnames(data),
+    must.include = c("date", "cases", "deaths")
+  )
+  # check for any NAs among data
+  checkmate::assert_data_frame(
+    data[, c("date", "cases", "deaths")],
+    any.missing = FALSE
+  )
+  checkmate::assert_number(smoothing_window, lower = 1, null.ok = TRUE)
+
+  stopifnot(
+    "`smoothing_window` must be an odd number greater than 0" =
+      (smoothing_window %% 2 != 0),
+    "`delay_density` must be a function with a single required argument,
+    and evaluating distribution density at a vector of values and returning a
+    numeric vector of the same length.
+    E.g. function(x) stats::dgamma(x = x, shape = 5, scale = 1)" =
+      (test_fn_req_args(delay_density) &&
+        test_fn_num_out(delay_density)) || is.null(delay_density)
+  )
+
+  # prepare a new dataframe with smoothed columns if requested
+  # all temporary operations are performed on df_temp,
+  # data is returned with only three new columns added, and no other changes
+  df_temp <- data
+  # smooth cases if requested
+  if (is.null(smoothing_window)) {
+    # set to 0 for internal use only --- see below
+    smoothing_window <- 0
+  } else {
+    # smooth data if requested
+    df_temp$cases <- stats::runmed(
+      data$cases,
+      k = smoothing_window,
+      endrule = "keep"
+    )
+
+    df_temp$deaths <- stats::runmed(
+      data$deaths,
+      k = smoothing_window,
+      endrule = "keep"
+    )
+  }
+
+  cases <- df_temp$cases
+
+  case_times <- as.numeric(df_temp$date - min(df_temp$date, na.rm = TRUE),
+    units = "days"
+  ) + 1
+
+  case_length <- length(case_times)
+
+  ##### prepare matrix for severity estimation ####
+  # when not correcting for delays, set estimated no. of known outcomes to cases
+  # this is to avoid if-else ladders
+  df_temp$estimated_outcomes <- df_temp$cases
+
+  # assign columns for severity estimate and intervals
+  severity_estimates <- matrix(
+    data = NA_real_, nrow = nrow(data), ncol = 3,
+    dimnames = list(NULL, sprintf("severity_%s", c("mean", "low", "high")))
+  )
+
+  # calculation of indices to modify
+  # start from the final index to be smoothed
+  # end at the first row after the burn-in number of rows (days)
+  # TODO: check if modifying start point is necessary for runmed "keep" endrule
+  indices <- seq(case_length - smoothing_window, burn_in + 1, -1)
+  if (!is.null(delay_density)) {
+    pmf_vals <- delay_density(seq(from = 0, to = nrow(data) - 1L))
+
+    df_temp[indices, "estimated_outcomes"] <- vapply(
+      X = indices,
+      FUN = function(x) {
+        delay_pmf_eval <- pmf_vals[case_times[seq_len(x - burn_in)]]
+        known_onsets_current <- cases[seq_len(x - burn_in)] *
+          rev(delay_pmf_eval)
+
+        # return rounded sum of known_onsets_current
+        round(sum(known_onsets_current, na.rm = TRUE))
```

Comment:
Thanks, a brief check suggests that these are redundant as they're calculated from columns in the data where no NAs are allowed, so the issue of what NAs indicate is pushed to prior data preparation steps.

Reply:
I've moved this calculation into an internal function in https://github.com/epiverse-trace/cfr/commit/70647b875cd9f986a527d919c36521a5aae3bfa4 which no longer removes `NAs`. Since this is an internal function it expects that the top-level functions check for `NA`s in the case time-series (and fail).

`NA`s in the case data would indicate missing data I suppose, such as a missed report; we had this discussing with regard to aggregating incidence2 data earlier as well, see #82.